### PR TITLE
Add VaultMinter automation contract

### PR DESCRIPTION
**Pull request by sunbreak1211**

### CODE REVIEW

File: test/funnels/automation/VaultMinter.t.sol [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        roles.setIlkAdmin(ILK, address(this));
+        roles.setRoleAction(ILK, MINTER_ROLE, address(vault), AllocatorVault.draw.selector, true);
+        roles.setRoleAction(ILK, MINTER_ROLE, address(vault), AllocatorVault.wipe.selector, true);
+        roles.setUserRole(ILK, address(minter), MINTER_ROLE, true);
+
+        // Set up keeper to mint and burn
+        minter.rely(FACILITATOR);
+        vm.startPrank(FACILITATOR);
+        minter.kiss(KEEPER);
+        vm.stopPrank();
```
Comment (oldchili) [inline test/funnels/automation/VaultMinter.t.sol]:
```suggestion
        vm.prank(FACILITATOR); minter.kiss(KEEPER);
```

File: src/funnels/automation/VaultMinter.sol [hunk 2]
```diff
@@ -0,0 +1,118 @@
+// SPDX-FileCopyrightText: © 2023 Dai Foundation <www.daifoundation.org>
```
Comment (oldchili) [inline src/funnels/automation/VaultMinter.sol:1]: Should we also add this new contract to the deployment scripts?

File: test/funnels/automation/VaultMinter.t.sol [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+        vm.warp(1 hours);
+    }
+
+    function testConstructor() public {
+        VaultMinter m = new VaultMinter(address(0xABC));
```
Comment (oldchili) [inline test/funnels/automation/VaultMinter.t.sol:70]: Can also check the existence of the `Rely` event here.

File: src/funnels/automation/VaultMinter.sol [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+// SPDX-License-Identifier: AGPL-3.0-or-later
```
Comment (oldchili) [inline src/funnels/automation/VaultMinter.sol:2]: We also need to add a paragraph in the README (can do next to the ConduitMover one.)

File: test/funnels/automation/VaultMinter.t.sol (hunk 3 above)
Comment (sunbreak1211) [reply test/funnels/automation/VaultMinter.t.sol:70]: Good catch, this is something we need to do in other contracts as well.

File: src/funnels/automation/VaultMinter.sol (hunk 2 above)
Comment (sunbreak1211) [reply src/funnels/automation/VaultMinter.sol:1]: Yes, had forgotten about this, working on it.

Comment (sunbreak1211) [reply src/funnels/automation/VaultMinter.sol:1]: Done.

File: test/funnels/automation/VaultMinter.t.sol (hunk 3 above)
Comment (sunbreak1211) [reply test/funnels/automation/VaultMinter.t.sol:70]: Done.

File: src/funnels/automation/VaultMinter.sol (hunk 4 above)
Comment (sunbreak1211) [reply src/funnels/automation/VaultMinter.sol:2]: Done. Let me know if you have something better to put here.

File: test/funnels/automation/VaultMinter.t.sol (hunk 1 above)
Comment (sunbreak1211) [reply test/funnels/automation/VaultMinter.t.sol]: Done.

File: test/funnels/automation/VaultMinter.t.sol (hunk 3 above)
Comment (sunbreak1211) [reply test/funnels/automation/VaultMinter.t.sol:70]: For the other contracts we should do a separate PR.

File: src/funnels/automation/VaultMinter.sol (hunk 4 above)
Comment (oldchili) [reply src/funnels/automation/VaultMinter.sol:2]: looks good.

File: test/integration/Deployment.t.sol [hunk 5]
```diff
@@ -278,13 +286,17 @@ contract DeploymentTest is DssTest {
...
+        assertEq(WardsLike(ilkInst.vaultMinter).wards(facilitator2), 1);
         assertEq(WardsLike(ilkInst.stableSwapper).wards(facilitator1), 1);
         assertEq(WardsLike(ilkInst.stableSwapper).wards(facilitator2), 1);
         assertEq(WardsLike(ilkInst.stableDepositorUniV3).wards(facilitator1), 1);
         assertEq(WardsLike(ilkInst.stableDepositorUniV3).wards(facilitator2), 1);
         assertEq(WardsLike(ilkInst.conduitMover).wards(facilitator1), 1);
         assertEq(WardsLike(ilkInst.conduitMover).wards(facilitator2), 1);
 
+        assertEq(ConduitMover(ilkInst.vaultMinter).buds(vaultMinterKeeper1), 1);
+        assertEq(ConduitMover(ilkInst.vaultMinter).buds(vaultMinterKeeper2), 1);
```
Comment (oldchili) [inline test/integration/Deployment.t.sol]:
```suggestion
        assertEq(VaultMinter(ilkInst.vaultMinter).buds(vaultMinterKeeper1), 1);
        assertEq(VaultMinter(ilkInst.vaultMinter).buds(vaultMinterKeeper2), 1);
```

File: src/funnels/automation/VaultMinter.sol [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        uint128  lot; // The amount to mint or burn every hop
+    }
+
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Kiss(address indexed usr);
+    event Diss(address indexed usr);
+    event SetConfig(int64 num, uint32 hop, uint128 lot);
+    event Mint(uint128 lot);
+    event Burn(uint128 lot);
```
Comment (oldchili) [inline src/funnels/automation/VaultMinter.sol]: Should we change these to Draw and Wipe now?

File: src/funnels/automation/VaultMinter.sol [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+contract VaultMinter {
+    mapping (address => uint256) public wards; // Admins
+    mapping (address => uint256) public buds;  // Whitelisted keepers
+    MinterConfig public config;                // Configuration for keepers
+
+    address public immutable vault; // The address of the vault contract
+
+    struct MinterConfig {
+        int64    num; // The remaining number of times that a mint or burn can be executed by keepers (> 0: mint, < 0: burn)
```
Comment (oldchili) [inline src/funnels/automation/VaultMinter.sol]:
```suggestion
        int64    num; // The remaining number of times that a mint or burn can be executed by keepers (> 0: draw, < 0: wipe)
```

File: src/funnels/automation/VaultMinter.sol [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        uint32   hop; // Cooldown period it has to wait between each action
+        uint32   zzz; // Timestamp of the last action
+        uint128  lot; // The amount to mint or burn every hop
```
Comment (oldchili) [inline src/funnels/automation/VaultMinter.sol]:
```suggestion
        uint128  lot; // The amount to draw or wipe every hop
```

File: test/funnels/automation/VaultMinter.t.sol (hunk 3 above)
Comment (telome) [reply test/funnels/automation/VaultMinter.t.sol:70]: Done here https://github.com/makerdao/dss-allocator/pull/71 (where I've also adopted your structure for the `not-keeper` revert tests in the other automation contracts, for consistency).

File: src/funnels/automation/VaultMinter.sol (hunk 7 above)
Comment (sunbreak1211) [reply src/funnels/automation/VaultMinter.sol]: Fix it and the a few other apparences of mint/burn terminology

File: src/funnels/automation/VaultMinter.sol (hunk 8 above)
Comment (sunbreak1211) [reply src/funnels/automation/VaultMinter.sol]: Done.

File: src/funnels/automation/VaultMinter.sol (hunk 6 above)
Comment (sunbreak1211) [reply src/funnels/automation/VaultMinter.sol]: Fixed.

**Review by oldchili — APPROVED** [review]
LGTM

**Review by telome — APPROVED** [review]
LGTM

**Pull request merged**